app/assets/javascripts/angular/services/annotation_refresher.js.erb

```diff
@@ -1,31 +1,73 @@
-angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "VariableInterpolator", function($http, VariableInterpolator) {
-  return function(graph, scope) {
-    var tags = graph.tags.map(function(e) {
-      if (e.name) {
-        var n = VariableInterpolator(e.name, scope.vars);
-        return n.split(",").map(function(s) { return s.trim(); });
-      } else {
-        return "";
+angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "$q", "$timeout", "VariableInterpolator", function($http, $q, $timeout, VariableInterpolator) {
+  var MINUTE = 1000*60;
+  var promises = {};
```

Comment:
dictionaries are unordered, so we would have to add some additional logic to manage which is the oldest.

The main purpose of this PR was to only make a single request. We could just delete the shared annotation server response as soon as it has been consumed, by putting it on a timer to clear after some (again, arbitrary) small amount of time within which all graphs should have finished loading, e.g. 10sec.